# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [345]:
# import numpy and pandas

import numpy as np 
import pandas as pd

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [346]:
# Run this code:

salaries = pd.read_csv('../Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [347]:
# Your code here:

print("·········")
print("Null columns:")
null_cols = salaries.isnull().sum()
print(null_cols[null_cols > 0])
print("·········")
print("Len:")
print(len(salaries.index))
print("·········")
print("Max Salary:")
print(max(salaries["Annual Salary"]))
print("·········")
display(salaries.head())

·········
Null columns:
Typical Hours    25161
Annual Salary     8022
Hourly Rate      25161
dtype: int64
·········
Len:
33183
·········
Max Salary:
300000.0
·········


Name                              Job Titles  \
0    AARON,  JEFFERY M                                SERGEANT   
1      AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)   
2  AARON,  KIMBERLEI R                CHIEF CONTRACT EXPEDITER   
3  ABAD JR,  VICENTE M                       CIVIL ENGINEER IV   
4    ABASCAL,  REECE E             TRAFFIC CONTROL AIDE-HOURLY   

         Department Full or Part-Time Salary or Hourly  Typical Hours  \
0            POLICE                 F           Salary            NaN   
1            POLICE                 F           Salary            NaN   
2  GENERAL SERVICES                 F           Salary            NaN   
3       WATER MGMNT                 F           Salary            NaN   
4              OEMC                 P           Hourly           20.0   

   Annual Salary  Hourly Rate  
0       101442.0          NaN  
1        94122.0          NaN  
2       101592.0          NaN  
3       110064.0          NaN  
4            NaN        19.86

We see from looking at the `head` function that there is quite a bit of missing data. Let's examine how much missing data is in each column. Produce this output in the cell below

In [348]:
# Your code here:

null_cols = salaries.isnull().sum()

print(null_cols[null_cols > 0])

Typical Hours    25161
Annual Salary     8022
Hourly Rate      25161
dtype: int64


Let's also look at the count of hourly vs. salaried employees. Write the code in the cell below

In [349]:
# Your code here:

print(salaries["Salary or Hourly"].value_counts())

Salary    25161
Hourly     8022
Name: Salary or Hourly, dtype: int64


What this information indicates is that the table contains information about two types of employees - salaried and hourly. Some columns apply only to one type of employee while other columns only apply to another kind. This is why there are so many missing values. Therefore, we will not do anything to handle the missing values.

There are different departments in the city. List all departments and the count of employees in each department.

In [350]:
# Your code here:

print(salaries["Department"].value_counts())

POLICE                   13414
FIRE                      4641
STREETS & SAN             2198
OEMC                      2102
WATER MGMNT               1879
AVIATION                  1629
TRANSPORTN                1140
PUBLIC LIBRARY            1015
GENERAL SERVICES           980
FAMILY & SUPPORT           615
FINANCE                    560
HEALTH                     488
CITY COUNCIL               411
LAW                        407
BUILDINGS                  269
COMMUNITY DEVELOPMENT      207
BUSINESS AFFAIRS           171
COPA                       116
BOARD OF ELECTION          107
DoIT                        99
PROCUREMENT                 92
INSPECTOR GEN               87
MAYOR'S OFFICE              85
CITY CLERK                  84
ANIMAL CONTRL               81
HUMAN RESOURCES             79
CULTURAL AFFAIRS            65
BUDGET & MGMT               46
ADMIN HEARNG                39
DISABILITIES                28
TREASURER                   22
HUMAN RELATIONS             16
BOARD OF

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [351]:
# Your code here:

from scipy.stats import ttest_1samp

h0: "μ = 30"
h1: "μ ≠ 30"

confidence_level = "95%"
α = 0.05

In [352]:
hourly_wages = []
for i in range(len(salaries["Hourly Rate"])):
    if salaries["Salary or Hourly"][i] == "Hourly":
        hourly_wages.append(salaries["Hourly Rate"][i])

In [353]:
ttest_1samp(hourly_wages, 30)

Ttest_1sampResult(statistic=20.6198057854942, pvalue=4.3230240486229894e-92)

In [354]:
pvalue=4.3230240486229894e-92

def ts_test_hypothesis(pvalue, α):
    if pvalue < α:
        return "We reject h0"
    if pvalue >= α:
        return "We don't reject h0"
        
print(ts_test_hypothesis(pvalue, α))

We reject h0


#### As we reject h0, we cannot say that the hourly wages are significantly similar to 30.

#### ---------------------------------

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are higher than last year's mean of $86000/year a year for all salaried employees. Test this one sided hypothesis using a 95% confidence interval.

Hint: A one tailed test has a p-value that is half of the two tailed p-value. If our hypothesis is greater than, then to reject, the test statistic must also be positive.

In [355]:
# Your code here:

lastyear_mu = 86000

h0: "μ > 86000"
h1: "μ <= 86000"

confidence_level = "95%"
α = 0.05

In [356]:
police_salaries = []

for i in range(len(salaries["Annual Salary"])):
    if salaries["Department"][i] == "POLICE":
        if salaries["Salary or Hourly"][i] == "Salary":
            if str(salaries["Annual Salary"][i]) != "nan":
                police_salaries.append(salaries["Annual Salary"][i])

In [357]:
ttest_1samp(police_salaries, 86000)

Ttest_1sampResult(statistic=3.081997005712994, pvalue=0.0020603403550965137)

In [358]:
pvalue=0.0020603403550965137

def os_test_hypothesis(pvalue, α):
    if pvalue/2 < α:
        return "We reject h0"
    if pvalue/2 > α:
        return "We don't reject h0"

print(os_test_hypothesis(pvalue, α))

We reject h0


#### As we reject h0, we cannot say that the salaries are greater than $86000/year"

#### ---------------------------------

Using the `crosstab` function, find the department that has the most hourly workers. 

In [359]:
salaries.head()

Name                              Job Titles  \
0    AARON,  JEFFERY M                                SERGEANT   
1      AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)   
2  AARON,  KIMBERLEI R                CHIEF CONTRACT EXPEDITER   
3  ABAD JR,  VICENTE M                       CIVIL ENGINEER IV   
4    ABASCAL,  REECE E             TRAFFIC CONTROL AIDE-HOURLY   

         Department Full or Part-Time Salary or Hourly  Typical Hours  \
0            POLICE                 F           Salary            NaN   
1            POLICE                 F           Salary            NaN   
2  GENERAL SERVICES                 F           Salary            NaN   
3       WATER MGMNT                 F           Salary            NaN   
4              OEMC                 P           Hourly           20.0   

   Annual Salary  Hourly Rate  
0       101442.0          NaN  
1        94122.0          NaN  
2       101592.0          NaN  
3       110064.0          NaN  
4            NaN        19.86

In [360]:
# Your code here:

deps_hourly_wages = pd.crosstab(salaries["Department"], salaries["Salary or Hourly"])

display(deps_hourly_wages.sort_values("Hourly", ascending = False).head())

Salary or Hourly  Hourly  Salary
Department                      
STREETS & SAN       1862     336
WATER MGMNT         1513     366
OEMC                1273     829
AVIATION            1082     547
GENERAL SERVICES     765     215

The workers from the department with the most hourly workers have complained that their hourly wage is less than $35/hour. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [361]:
# Your code here:

m = 35

h0: "μ > 35"
h1: "μ <= 35"

confidence_level = "95%"
α = 0.05

In [362]:
streets_san_wages = []

for i in range(len(salaries["Hourly Rate"])):
    if salaries["Department"][i] == "STREETS & SAN":
        if salaries["Salary or Hourly"][i] == "Hourly":
            streets_san_wages.append(salaries["Hourly Rate"][i])

In [363]:
ttest_1samp(streets_san_wages, m)

Ttest_1sampResult(statistic=-9.567447887848152, pvalue=3.3378530564707717e-21)

In [364]:
pvalue=3.3378530564707717e-21

def os_test_hypothesis(pvalue, α):
    if pvalue/2 < α:
        return "We reject h0"
    if pvalue/2 > α:
        return "We don't reject h0"

print(os_test_hypothesis(pvalue, α))

We reject h0


#### As we reject the null hypothesis, we cannot say that the hourly wages of the STREETS & SAN workers are greater than $35

#### -------------------

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [365]:
# Your code here:

hourly_wages_df = salaries[salaries["Hourly Rate"].notnull()]

hourly_wages_df.head()

Name                   Job Titles        Department  \
4       ABASCAL,  REECE E  TRAFFIC CONTROL AIDE-HOURLY              OEMC   
6   ABBATACOLA,  ROBERT J          ELECTRICAL MECHANIC          AVIATION   
7      ABBATE,  JOSEPH L       POOL MOTOR TRUCK DRIVER     STREETS & SAN   
10       ABBOTT,  BETTY L           FOSTER GRANDPARENT  FAMILY & SUPPORT   
18   ABDULLAH,  LAKENYA N               CROSSING GUARD              OEMC   

   Full or Part-Time Salary or Hourly  Typical Hours  Annual Salary  \
4                  P           Hourly           20.0            NaN   
6                  F           Hourly           40.0            NaN   
7                  F           Hourly           40.0            NaN   
10                 P           Hourly           20.0            NaN   
18                 P           Hourly           20.0            NaN   

    Hourly Rate  
4         19.86  
6         46.10  
7         35.60  
10         2.65  
18        17.68

In [366]:
from scipy import stats

confidence_level = 0.95
α = 0.05
degrees_of_freedom = len(hourly_wages_df["Hourly Rate"]) - 1
loc = np.mean(hourly_wages_df["Hourly Rate"])
SE = stats.sem(hourly_wages_df["Hourly Rate"])

Now construct the 95% confidence interval for all salaried employeed in the police in the cell below.

In [367]:
# Your code here:

confidence_interval = stats.t.interval(confidence_level, degrees_of_freedom, loc, SE)

print(confidence_interval)

(32.52345834488425, 33.05365708767623)


#### With a 95% level of confidence, I can say that the mean of the sample is between 32.52345834488425 and 33.05365708767623.

#### -----------------------------

# Bonus Challenge - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [368]:
display(salaries.head())

Name                              Job Titles  \
0    AARON,  JEFFERY M                                SERGEANT   
1      AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)   
2  AARON,  KIMBERLEI R                CHIEF CONTRACT EXPEDITER   
3  ABAD JR,  VICENTE M                       CIVIL ENGINEER IV   
4    ABASCAL,  REECE E             TRAFFIC CONTROL AIDE-HOURLY   

         Department Full or Part-Time Salary or Hourly  Typical Hours  \
0            POLICE                 F           Salary            NaN   
1            POLICE                 F           Salary            NaN   
2  GENERAL SERVICES                 F           Salary            NaN   
3       WATER MGMNT                 F           Salary            NaN   
4              OEMC                 P           Hourly           20.0   

   Annual Salary  Hourly Rate  
0       101442.0          NaN  
1        94122.0          NaN  
2       101592.0          NaN  
3       110064.0          NaN  
4            NaN        19.86

In [369]:
print(salaries["Salary or Hourly"].value_counts())

Salary    25161
Hourly     8022
Name: Salary or Hourly, dtype: int64


In [370]:
# Your code here:

from statsmodels.stats.proportion import proportions_ztest

p = 0.25

h0: "μ = p"
h1: "μ ≠ p" 

count = 8022
nobs = 8022 + 25161
#count = 1
#nobs = 4
α = 0.05

proportions_ztest(count, nobs, p)

(-3.5099964213703005, 0.0004481127249057967)

In [371]:
pvalue = 0.0004481127249057967

def ts_test_hypothesis(pvalue, α):
    if pvalue < α:
        return "We reject h0"
    if pvalue >= α:
        return "We don't reject h0"

print(ts_test_hypothesis(pvalue, α))

We reject h0


### As we reject our null hypothesis, we cannot say that the proportion of hourly workers in the City of Chicago is significantly similar to 25%

In [372]:
print(8022/(8022 + 25161))

0.24175029382515142


In [373]:
# Me sorprende que me diga que rechace h0, si la proporción está tan cerca de 0.25